distributed/comm/asyncio_tcp.py

```diff
@@ -0,0 +1,782 @@
+from .registry import Backend
+from .utils import ensure_concrete_host, from_frames, to_frames
+
+logger = logging.getLogger(__name__)
+
+
+_COMM_CLOSED = object()
+
+
+class DaskCommProtocol(asyncio.BufferedProtocol):
+    def __init__(self, on_connection=None, min_read_size=128 * 1024):
+        super().__init__()
+        self.on_connection = on_connection
+        self._exception = None
+        self._queue = asyncio.Queue()
+        self._transport = None
+        self._paused = False
+        self._drain_waiter = None
+        self._loop = asyncio.get_running_loop()
+        self._is_closed = self._loop.create_future()
+
+        # Per-message state
+        self._using_default_buffer = True
+
+        self._default_buffer = memoryview(bytearray(min_read_size))
+        self._default_len = min_read_size
+        self._default_start = 0
+        self._default_end = 0
+
+        self._nframes = None
+        self._frame_lengths = None
+        self._frames = None
+        self._frame_index = None
+        self._frame_nbytes_needed = 0
+
+    @property
+    def local_addr(self):
+        if self._transport is None:
+            return "<closed>"
+        try:
+            host, port = self._transport.get_extra_info("socket").getsockname()[:2]
+            return unparse_host_port(host, port)
+        except Exception:
+            return "<closed>"
+
+    @property
+    def peer_addr(self):
+        if self._transport is None:
+            return "<closed>"
+        peername = self._transport.get_extra_info("peername")
+        if peername is not None:
+            return unparse_host_port(*peername[:2])
+        return "<unknown>"
+
+    @property
+    def is_closed(self):
+        return self._transport is None
+
+    def _abort(self):
+        if self._transport is not None:
+            self._transport, transport = None, self._transport
+            transport.abort()
+
+    def _close_from_finalizer(self, comm_repr):
+        if self._transport is not None:
+            logger.warning(f"Closing dangling comm `{comm_repr}`")
+            try:
+                self._abort()
+            except RuntimeError:
+                # This happens if the event loop is already closed
+                pass
+
+    async def _close(self):
+        if self._transport is not None:
+            self._transport, transport = None, self._transport
+            transport.close()
+        await self._is_closed
+
+    def connection_made(self, transport):
+        if type(transport) is asyncio.selector_events._SelectorSocketTransport:
+            transport = _ZeroCopyWriter(transport)
+        self._transport = transport
+        if self.on_connection is not None:
+            self.on_connection(self)
+
+    def get_buffer(self, sizehint):
+        if self._frames is None or self._frame_nbytes_needed < self._default_len:
+            self._using_default_buffer = True
+            return self._default_buffer[self._default_end :]
```

Comment:
Since we memcpy shift the default buffer at the end of every parse loop, at this point `_default_end` can be at most `7` (not quite enough to read another uint8). I'll add a check on init to ensure `_default_len >=  8`, but otherwise we should be fine here.